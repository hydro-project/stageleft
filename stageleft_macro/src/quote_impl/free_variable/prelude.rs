const PRELUDE: &[&str] = &["Err", "None", "Ok", "Some"];

pub fn is_prelude(ident: &syn::Ident) -> bool {
    let ident_str = ident.to_string();
    // Linear search is fast enough for four items.
    PRELUDE.contains(&ident_str.as_str())
}
