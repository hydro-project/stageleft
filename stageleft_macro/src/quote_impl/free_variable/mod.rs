use std::collections::{BTreeSet, HashSet};

mod prelude;
use prelude::is_prelude;

#[derive(Debug)]
pub struct ScopeStack {
    scopes: Vec<(HashSet<String>, HashSet<String>)>,
}

impl Default for ScopeStack {
    fn default() -> Self {
        ScopeStack {
            scopes: vec![(HashSet::new(), HashSet::new())],
        }
    }
}

impl ScopeStack {
    pub fn push(&mut self) {
        self.scopes.push((HashSet::new(), HashSet::new()));
    }

    pub fn pop(&mut self) {
        self.scopes.pop();
    }

    pub fn insert_term(&mut self, ident: syn::Ident) {
        self.scopes
            .last_mut()
            .expect("Scope stack should not be empty")
            .0
            .insert(ident.to_string());
    }

    pub fn insert_type(&mut self, ident: syn::Ident) {
        self.scopes
            .last_mut()
            .expect("Scope stack should not be empty")
            .1
            .insert(ident.to_string());
    }

    pub fn contains_term(&self, ident: &syn::Ident) -> bool {
        let ident = ident.to_string();
        self.scopes
            .iter()
            .rev()
            .any(|scope| scope.0.contains(&ident))
    }

    pub fn contains_type(&self, ident: &syn::Ident) -> bool {
        let ident = ident.to_string();
        self.scopes
            .iter()
            .rev()
            .any(|scope| scope.1.contains(&ident))
    }
}

#[derive(Default)]
pub struct FreeVariableVisitor {
    pub free_variables: BTreeSet<syn::Ident>,
    pub current_scope: ScopeStack,
}

impl<'a> syn::visit::Visit<'a> for FreeVariableVisitor {
    fn visit_expr_closure(&mut self, i: &syn::ExprClosure) {
        self.current_scope.push();
        i.inputs.iter().for_each(|input| {
            self.visit_pat(input);
        });

        syn::visit::visit_expr_closure(self, i);

        self.current_scope.pop();
    }

    fn visit_item_fn(&mut self, i: &syn::ItemFn) {
        self.current_scope.push();
        syn::visit::visit_item_fn(self, i);
        self.current_scope.pop();
    }

    fn visit_generic_param(&mut self, i: &syn::GenericParam) {
        match i {
            syn::GenericParam::Type(type_param) => {
                self.current_scope.insert_type(type_param.ident.clone());
            }
            syn::GenericParam::Lifetime(lifetime_param) => {
                self.current_scope
                    .insert_type(lifetime_param.lifetime.ident.clone());
            }
            syn::GenericParam::Const(const_param) => {
                self.current_scope.insert_type(const_param.ident.clone());
            }
        }
    }

    fn visit_block(&mut self, i: &syn::Block) {
        self.current_scope.push();
        syn::visit::visit_block(self, i);
        self.current_scope.pop();
    }

    fn visit_local(&mut self, i: &syn::Local) {
        i.init.iter().for_each(|init| {
            syn::visit::visit_local_init(self, init);
        });
        match &i.pat {
            syn::Pat::Ident(pat_ident) => {
                self.current_scope.insert_term(pat_ident.ident.clone());
            }
            syn::Pat::Type(pat_type) => {
                self.visit_pat(&pat_type.pat);
            }
            syn::Pat::Wild(_) => {
                // Do nothing
            }
            syn::Pat::Tuple(pat_tuple) => {
                for el in &pat_tuple.elems {
                    self.visit_pat(el);
                }
            }
            _ => panic!("Local variables must be identifiers, got {:?}", i.pat),
        }
    }

    fn visit_ident(&mut self, i: &proc_macro2::Ident) {
        if !self.current_scope.contains_term(i) {
            self.free_variables.insert(i.clone());
        }
    }

    fn visit_lifetime(&mut self, i: &syn::Lifetime) {
        if !self.current_scope.contains_type(&i.ident) {
            self.free_variables.insert(i.ident.clone());
        }
    }

    fn visit_path(&mut self, i: &syn::Path) {
        if i.leading_colon.is_none() && !is_prelude(&i.segments.first().unwrap().ident) {
            let one_segment = i.segments.len() == 1;
            let node = i.segments.first().unwrap();
            if one_segment && !self.current_scope.contains_term(&node.ident) {
                self.free_variables.insert(node.ident.clone());
            }
        }

        for node in i.segments.iter() {
            self.visit_path_arguments(&node.arguments);
        }
    }

    fn visit_arm(&mut self, i: &syn::Arm) {
        self.current_scope.push();
        syn::visit::visit_arm(self, i);
        self.current_scope.pop();
    }

    fn visit_field_pat(&mut self, i: &syn::FieldPat) {
        for it in &i.attrs {
            self.visit_attribute(it);
        }
        self.visit_pat(&i.pat);
    }

    fn visit_pat_ident(&mut self, i: &syn::PatIdent) {
        self.current_scope.insert_term(i.ident.clone());
    }

    fn visit_expr_call(&mut self, i: &syn::ExprCall) {
        if let syn::Expr::Path(path) = &*i.func {
            if path.path.segments.len() == 1 {
                // skip, don't emit a free variable, assume that it's a
                // fn call and not a closure call; for closure calls, we
                // require the user to manually capture it with a `let` binding
            } else {
                syn::visit::visit_expr(self, &i.func);
            }
        } else {
            syn::visit::visit_expr(self, &i.func);
        }

        for arg in &i.args {
            self.visit_expr(arg);
        }
    }

    fn visit_expr_method_call(&mut self, i: &syn::ExprMethodCall) {
        syn::visit::visit_expr(self, &i.receiver);
        for arg in &i.args {
            self.visit_expr(arg);
        }
    }

    fn visit_type(&mut self, _: &syn::Type) {}

    fn visit_expr_struct(&mut self, node: &syn::ExprStruct) {
        for it in &node.attrs {
            self.visit_attribute(it);
        }
        if let Some(it) = &node.qself {
            self.visit_qself(it);
        }
        // No need to capture the struct path
        // self.visit_path(&node.path);
        for el in syn::punctuated::Punctuated::pairs(&node.fields) {
            let it = el.into_value();
            self.visit_expr(&it.expr);
        }
        if let Some(it) = &node.rest {
            self.visit_expr(it);
        }
    }

    fn visit_expr_field(&mut self, i: &syn::ExprField) {
        self.visit_expr(&i.base);
    }

    fn visit_macro(&mut self, i: &syn::Macro) {
        // TODO(shadaj): emit a warning if our guess at parsing fails
        match i.delimiter {
            syn::MacroDelimiter::Paren(_binding_0) => {
                i.parse_body_with(
                    syn::punctuated::Punctuated::<syn::Expr, syn::Token![,]>::parse_terminated,
                )
                .ok()
                .into_iter()
                .for_each(|exprs| {
                    for arg in &exprs {
                        self.visit_expr(arg);
                    }
                });
            }
            syn::MacroDelimiter::Brace(_binding_0) => {
                i.parse_body_with(syn::Block::parse_within)
                    .ok()
                    .into_iter()
                    .for_each(|stmts| {
                        for stmt in &stmts {
                            self.visit_stmt(stmt);
                        }
                    });
            }
            syn::MacroDelimiter::Bracket(_binding_0) => {
                i.parse_body_with(
                    syn::punctuated::Punctuated::<syn::Expr, syn::Token![,]>::parse_terminated,
                )
                .ok()
                .into_iter()
                .for_each(|exprs| {
                    for arg in &exprs {
                        self.visit_expr(arg);
                    }
                });
            }
        }
    }
}
