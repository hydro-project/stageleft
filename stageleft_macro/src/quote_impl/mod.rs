use proc_macro2::{Span, TokenStream};
use quote::{ToTokens, quote};
use syn::visit_mut::VisitMut;

use self::free_variable::FreeVariableVisitor;

mod free_variable;

pub fn q_impl(root: TokenStream, toks: proc_macro2::TokenStream) -> TokenStream {
    let mut visitor = FreeVariableVisitor::default();
    let rewritten_toks = if let Ok(mut expr) = syn::parse2::<syn::Expr>(toks.clone()) {
        visitor.visit_expr_mut(&mut expr);
        expr.into_token_stream()
    } else {
        toks
    };

    let unitialized_free_variables = visitor.free_variables.iter().map(|i| {
        let i_without_span = syn::Ident::new(&i.to_string(), Span::call_site());

        let ident_shadow_str = format!("{}__free", i);
        let i_shadow_ident = syn::Ident::new(&ident_shadow_str, Span::call_site());

        quote!(
            #[allow(unused, non_upper_case_globals, non_snake_case)]
            let #i_shadow_ident = {
                let _out = ::#root::runtime_support::FreeVariableWithContext::uninitialized(&#i_without_span, __stageleft_ctx);
                __output.captures.push(::#root::internal::Capture {
                    ident: #ident_shadow_str,
                    tokens: ::#root::runtime_support::FreeVariableWithContext::to_tokens(#i_without_span, __stageleft_ctx),
                });
                _out
            };
        )
    });

    let uninit_forgets = visitor.free_variables.iter().map(|i| {
        let i_shadow_ident = syn::Ident::new(&format!("{}__free", i), Span::call_site());

        quote!(
            #[allow(unused, non_upper_case_globals, non_snake_case)]
            ::std::mem::forget(#i_shadow_ident);
        )
    });

    // necessary to ensure proper hover in Rust Analyzer
    let expr_without_spans = if std::env::var("RUST_ANALYZER_INTERNALS_DO_NOT_USE").is_ok() {
        // for unknown reasons, Rust Analyzer completions break if we emit the input as a string as well :(
        "".to_string()
    } else {
        rewritten_toks.to_string()
    };

    quote!({
        move |__stageleft_ctx: &_, __output: &mut ::#root::internal::QuotedOutput| {
            #(#unitialized_free_variables;)*

            if true {
                __output.module_path = module_path!();
                __output.crate_name = option_env!("STAGELEFT_FINAL_CRATE_NAME").unwrap_or(env!("CARGO_PKG_NAME"));
                __output.tokens = #expr_without_spans;

                #(#uninit_forgets;)*
                unsafe {
                    return ::std::mem::MaybeUninit::uninit().assume_init();
                }
            }

            #[allow(unreachable_code, unused_qualifications)]
            {
                #rewritten_toks
            }
        }
    })
}

#[cfg(test)]
mod tests {
    use quote::quote;
    use syn::parse_quote;

    use super::q_impl;

    macro_rules! test_quote {
        ($program:expr) => {
            let quoted_tokens = q_impl(quote!(stageleft), parse_quote!($program));
            let wrapped: syn::File = parse_quote! {
                fn main() {
                    #quoted_tokens
                }
            };

            insta::with_settings!({snapshot_suffix => "macro_tokens"}, {
                insta::assert_snapshot!(
                    prettyplease::unparse(&wrapped)
                );
            });
        };
    }

    #[test]
    fn test_simple() {
        test_quote! {
            1 + 2
        }
    }

    #[test]
    fn test_capture_local() {
        test_quote! {
            x + 2
        }
    }

    #[test]
    fn test_capture_copy_local() {
        test_quote! {
            (x + 2) + (x + 2)
        }
    }

    #[test]
    fn test_capture_copy_local_block() {
        test_quote! {{
            let _ = x + 2;
            let _ = x + 2;
        }}
    }

    #[test]
    fn test_capture_copy_local_block_let() {
        test_quote! {{
            let x = x + 2;
            let _ = x + 2;
        }}
    }

    #[test]
    fn test_capture_local_mut() {
        test_quote! {
            x += 2
        }
    }

    #[test]
    fn test_non_capture_local() {
        test_quote! {
            {
                let x = 1;
                x + 2
            }
        }
    }

    #[test]
    fn test_capture_in_macro() {
        test_quote! {
            dbg!(x)
        }
    }

    #[test]
    fn test_non_capture_struct_creation() {
        test_quote! {
            Foo { x: 1 }
        }
    }

    #[test]
    fn test_non_capture_enum_creation() {
        test_quote! {
            Foo::Bar { x: 1 }
        }
    }

    #[test]
    fn test_prelude_enum_variants() {
        test_quote! {
            Some(1)
        }

        test_quote! {
            None
        }

        test_quote! {
            Ok(1)
        }

        test_quote! {
            Err(1)
        }
    }
}
