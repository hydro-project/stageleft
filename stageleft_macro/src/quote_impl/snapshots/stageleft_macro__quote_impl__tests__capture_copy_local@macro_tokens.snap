---
source: stageleft_macro/src/quote_impl/mod.rs
expression: "prettyplease :: unparse(& wrapped)"
---
fn main() {
    {
        move |
            __stageleft_ctx: &_,
            set_mod: &mut String,
            set_crate_name: &mut &'static str,
            set_tokens: &mut &'static str,
            _vec_to_set: &mut stageleft::internal::CaptureVec,
            run: bool|
        {
            #[allow(unused, non_upper_case_globals, non_snake_case)]
            let x = {
                let _out = ::stageleft::runtime_support::FreeVariableWithContext::uninitialized(
                    &x,
                    __stageleft_ctx,
                );
                _vec_to_set
                    .push((
                        "x".to_string(),
                        ::stageleft::runtime_support::FreeVariableWithContext::to_tokens(
                            x,
                            __stageleft_ctx,
                        ),
                    ));
                _out
            };
            if !run {
                *set_mod = module_path!().to_string();
                *set_crate_name = option_env!("STAGELEFT_FINAL_CRATE_NAME")
                    .unwrap_or(env!("CARGO_PKG_NAME"));
                *set_tokens = "(x + 2) + (x + 2)";
                #[allow(unused, non_upper_case_globals, non_snake_case)]
                ::std::mem::forget(x);
                unsafe {
                    return ::std::mem::MaybeUninit::uninit().assume_init();
                }
            }
            #[allow(unreachable_code, unused_qualifications)] { (x + 2) + (x + 2) }
        }
    }
}
